File: dynamic_programming_1D/longestPalindromicSubstring_Me.js
/*
  Leetcode: 5. Longest Palindromic Substring (medium)

  Given a string s, return the longest palindromic substring in s.


  Example 1:
    Input: s = "babad"
    Output: "bab"
    Explanation: "aba" is also a valid answer.

  Example 2:
    Input: s = "cbbd"
    Output: "bb"
  

  Constraints:
    - 1 <= s.length <= 1000
    - s consist of only digits and English letters.

  Solution by myself:
    using two pointer 

  Leetcode submission:
    #1
      Runtime: 600 ms, beats 29.38%
      Memory: 47.8 MB, beats 46.31%
    #2 (break the loop when j === s.length -1 when palindrome found)
      Runtime: 518 ms, beats 33.48%
      Memory: 47.9 MB, beats 43.97%
    #3 (set variable minJ to store min j index when palindrome found, the is j is <= minJ then reset continue)
      Runtime: 535 ms, beats 32.25%
      Memory: 47.5 MB, beats 47.54%
    #4 if substr.length < result.length reset continue
        if substr.length < result.length && j === s.length - 1 break
      Runtime: 445 ms, beats 37.15%
      Memory: 47.9 MB, beats 43.97%
      
*/

function longestPalindrome(s) {
   if (s.length === 1) return s;

   function isPalindrome(str) {
      let j = str.length - 1;
      for (let i = 0; i < Math.floor(str.length / 2); i++) {

         if (str[i] !== str[j]) {
            return false;
         }
         if (j < i) {
            // reach mid
            break;
         }
         j--;
      }
      return true;
   }

   // let possible = []; // possible string tobe palindrome
   // let 
   let result = '';
   let j = s.length - 1;
   let minJ = 0; // min j index when new palindrom length > result length
   let i = 0;
   x = 0;
   while (i < s.length) {
      console.log(s.slice(i, j + 1), i, j);
      const substr = s.slice(i, j + 1);
      // #4 [start]
      if (substr.length <= result.length || j < i || j <= minJ) {
         if (j === s.length - 1) break;
         console.log('reset');
         i++;
         j = s.length - 1;
         continue;
      }
      // #4 [end]

      // #3 [start]
      if (j < i || j <= minJ) {
         console.log('reset');
         i++;
         j = s.length - 1;
         continue;
      }
      // #3 [end]
      if (s[i] !== s[j]) {
         j--;
      } else {
         const isP = isPalindrome(substr)
         if (isP) {
            // #3 [start]
            if (substr.length > result.length) {
               result = substr;
               console.log('FOUND PALINDROME', substr);
               minJ = j;
            }
            // #3 [end]
            // #2 [start]
            // when j is s.length - 1 then is the longest
            if (j === s.length - 1) {
               break;
            }
            // #2 [end]
            j = s.length - 1;
            i++;
            console.log('true palindrome');
            continue;
         }
         j--;
         continue;
      }

   }
   return result;
}
const s1 = 'babad';
const s2 = 'cbbd';
const s3 = 'aacabdkacaa'; // aca
const c1 = 'aba';
const c2 = 'bbxxbb';
const c3 = 'lasdjfiejfifjababababababababababababasdfrasdfiejfiejfiejfiljfsdklfja;sldfjseifjisefjlsidjfisjf';
const s4 = "abacab"; // bacab
const s5 = 'ac'; // a
const start = Date.now();
console.log('RESULT: ', longestPalindrome(c3));
console.log('RUNTIME: ', Date.now() - start);
function isPalindrome(str) {
   let j = str.length - 1;
   for (let i = 0; i < Math.floor(str.length / 2); i++) {

      if (str[i] !== str[j]) {
         return false;
      }
      if (j < i) {
         // reach mid
         break;
      }
      j--;
   }
   return true;
}
 // console.log('is palindrome: ', isPalindrome('a'));

File: dynamic_programming_1D/longestPalindromicSubstring_Me.py
import datetime
'''
  Leetcode: 5. Longest Palindromic Substring (medium)

  Given a string s, return the longest palindromic substring in s.


  Example 1:
    Input: s = "babad"
    Output: "bab"
    Explanation: "aba" is also a valid answer.

  Example 2:
    Input: s = "cbbd"
    Output: "bb"
  

  Constraints:
    - 1 <= s.length <= 1000
    - s consist of only digits and English letters.

  Solution by myself:

  Leetcode submission:
   #1
      Runtime: 324 ms, beats 94.39%
      Memory: 16.6 MB, beats 24.65%
   #2 remove if r == len(s): return result and loop until len(s) - 1
      Runtime: 343 ms, beats 94.6%
      Memory: 16.4 MB, beats 34.49%
      
'''



def longestPalindrome(s):
   if len(s) == 1: return s

   result = s[0]
   for i in range(len(s)-1):
      c = s[i]

      l = i - 1
      r = i + 1
      set = {s[i]}

      if (s[l] != c or l == -1) and c == s[r]:
         result = s[i:r+1] if len(s[i:r+1]) > len(result) else result
         set.add(s[r])
         r += 1
      
      while l >= -1 and r < len(s):
         if len(result) == len(s): return result

         if i >= 20 and i <= 28:
            print('i', i, 'l', l, 'r', r)
            print('result', result, 'substr', s[l: r+1])
            if (s[r] == s[l+1]):
               print('s[r] == s[l+1]', s[r], s[l+1])
               print('s[l] == s[r]', s[l], s[r])

         if l >= 0 and s[l] == s[r]:
            print('BOTH EQUAL')
            result = s[l:r+1] if len(s[l:r+1]) > len(result) else result
            set.add(s[l]); set.add(s[r])
            l = l - 1
            r +=1
         elif s[r] == s[l+1]:
            print('ELIF', set)
            if len(set) == 1:
               result = s[l+1:r+1] if len(s[l+1:r+1]) > len(result) else result 
               r += 1
            else:
               break
         else:
            break
   
   return result
s9 = 'astxtsxas' # expect: stxts
s10 = 'aaaa' # expect: aaaa, output: aaa
s11 = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' # length = 47
s12 = 'aaabaaaa' # expect: aaabaaa, output: aaabaaaa
s13 = 'xxxxccxxxxxx' # expect: xxxxccxxxx
s14 = 'xxxxxxccxxxx' # expect: xxxxccxxxx
s15 = 'xxxxxxcccxxxx' # expect: xxxxcccxxxx
s16 = 'xxxxxxcacxxxx' # expect: xxxxcacxxxx
s17 = 'caaaaa' # expect: aaaaa, output: aa
s18 = 'aaaac' # expect: aaaa
s19 = 'acaa' # expect: aca
s20 = 'abbcccbbbcaaccbababcbcabca' # expect: bbcccbb, output: bbcccbbb
s21 = 'aaaabbbbbbbbbbccccccccccddddddddddeeeeeeeeeeffffffffffgggggggggghhhhhhhhhhiiiiiiiiiijjjjjjjjjjkkkkkkkkkkllllllllllmmmmmmmmmmnnnnnnnnnnooooooooooppppppppppqqqqqqqqqqrrrrrrrrrrssssssssssttttttttttuuuuuuuuuuvvvvvvvvvvwwwwwwwwwwxxxxxxxxxxyyyyyyyyyyzzzzzzzzzzyyyyyyyyyyxxxxxxxxxxwwwwwwwwwwvvvvvvvvvvuuuuuuuuuuttttttttttssssssssssrrrrrrrrrrqqqqqqqqqqppppppppppoooooooooonnnnnnnnnnmmmmmmmmmmllllllllllkkkkkkkkkkjjjjjjjjjjiiiiiiiiiihhhhhhhhhhggggggggggffffffffffeeeeeeeeeeddddddddddccccccccccbbbbbbbbbbaaaaaaaabbbbbbbbbbccccccccccddddddddddeeeeeeeeeeffffffffffgggggggggghhhhhhhhhhiiiiiiiiiijjjjjjjjjjkkkkkkkkkkllllllllllmmmmmmmmmmnnnnnnnnnnooooooooooppppppppppqqqqqqqqqqrrrrrrrrrrssssssssssttttttttttuuuuuuuuuuvvvvvvvvvvwwwwwwwwwwxxxxxxxxxxyyyyyyyyyyzzzzzzzzzzyyyyyyyyyyxxxxxxxxxxwwwwwwwwwwvvvvvvvvvvuuuuuuuuuuttttttttttssssssssssrrrrrrrrrrqqqqqqqqqqppppppppppoooooooooonnnnnnnnnnmmmmmmmmmmllllllllllkkkkkkkkkkjjjjjjjjjjiiiiiiiiiihhhhhhhhhhggggggggggffffffffffeeeeeeeeeeddddddddddccccccccccbbbbbbbbbbaaaa' # expect: same
s22 = 'ccccccccccbbbbbbbbbbaaaaaaaabbbbbbbbbbcccccccccc'
s23 = 'bb'

start = datetime.datetime.now().timestamp()
print('Result: ', longestPalindrome('cbbd'))
# s = 'aaac'
# print('s[-1:4]', s[-1:4])
print('Runtime: ', datetime.datetime.now().timestamp() - start)